src/object.c

```diff
@@ -377,63 +377,120 @@ void decrRefCount(robj *o) {
-        dontNeedMemory(((dict*)o->ptr)->ht[1].table);
-        break;
-    case OBJ_ENCODING_ZIPLIST:
-        dontNeedMemory(o->ptr);
-        break;
-    default:
-        serverPanic("Unknown hash encoding type");
+void dismissZsetObject(robj *o, size_t dump_size) {
+    if (o->encoding == OBJ_ENCODING_SKIPLIST) {
+        zset *zs = o->ptr;
+        zskiplist *zsl = zs->zsl;
+        /* We iterate all members only when average member size
+         * is more than page size. */
+        if (dump_size / zsl->length >= server.page_size) {
+            /* Dismiss members memory. */
+            zskiplistNode *zn = zsl->tail;
+            while (zn != NULL) {
+                dismissSds(zn->ele);
+                zn = zn->backward;
+            }
+        }
+
+        /* Dismiss hash table memory. */
+        dismissMemory(zs->dict->ht[0].table);
+        dismissMemory(zs->dict->ht[1].table);
+    } else if (o->encoding == OBJ_ENCODING_ZIPLIST) {
+        dismissMemory(o->ptr);
+    }
+}
+
+void dismissHashObject(robj *o, size_t dump_size) {
+    if (o->encoding == OBJ_ENCODING_HT) {
+        dict *d = o->ptr;
+        /* We iterate all field/values only when average field/values
+         * size is more than page size. */
+        if (dump_size / dictSize(d) >= server.page_size) {
+            /* Only dismiss values memory since the field size usually
+             * is small. */
+            dictEntry *de;
+            dictIterator *di = dictGetIterator(d);
+            while((de = dictNext(di)) != NULL) {
+                dismissSds(dictGetKey(de));
+                dismissSds(dictGetVal(de));
+            }
+            dictReleaseIterator(di);
+        }
+
+        /* Dismiss hash table memory. */
+        dismissMemory(d->ht[0].table);
+        dismissMemory(d->ht[1].table);
+    } else if (o->encoding == OBJ_ENCODING_ZIPLIST) {
+        dismissMemory(o->ptr);
     }
 }
 
-void dontNeedObject(robj *o) {
+/* Because it may cost much time to iterate all node/field/members of complex
+ * data type, we decide to iterate and dismiss them only when approximate
+ * average node/field/member size is more than page size of OS. 'dump_size'
+ * is the size of serialized key values. This method is not accurate, but it
+ * can avoid unnecessary iteration for complex data type. */
+void dismissObject(robj *o, size_t dump_size) {
     if (o->refcount != 1) return;
     switch(o->type) {
-        case OBJ_STRING: dontNeedStringObject(o); break;
-        case OBJ_SET: dontNeedSetObject(o); break;
-        case OBJ_ZSET: dontNeedZsetObject(o); break;
-        case OBJ_HASH: dontNeedHashObject(o); break;
+        case OBJ_STRING: dismissStringObject(o); break;
+        case OBJ_LIST: dismissListObject(o, dump_size); break;
+        case OBJ_SET: dismissSetObject(o, dump_size); break;
+        case OBJ_ZSET: dismissZsetObject(o, dump_size); break;
+        case OBJ_HASH: dismissHashObject(o, dump_size); break;
```

Comment:
what about streams? since they're listpack encoded, dismissing them could be similar to quicklist.
i.e. iterate on the rax, and release whole listpacks. and do that only if the average element size of the rax is big.

Reply:
I only used rax tree, but not familiar with it, at first glance, i find rax iterator is not safe, we may still use listpacks  data when iterating rax. i may have a look at `raxRecursiveFree`, any advise?

Reply:
you just need to use raxIterator.. have a look at `defragRadixTree`

Reply:
actually that function has some complications that can mislead you.
basically something like that:
```c
    raxIterator ri;
    raxStart(&ri,rax);
    raxSeek(&ri,"^",NULL,0);
    while (raxNext(&ri)) {
            madvise(ri.data);
    }
    raxStop(&ri);
```

and the judgement about iterating or not should use `raxSize`

Reply:
i am afraid we can't, we may access `ri.data` when calling `raxNext`

Reply:
are you saying that from code review, or did you test it and it crashed?

are you sure you're not confusing it with `it->node->data`?
each node has an encoded payload with pointers to the children and the fragments of the keys, as well as the pointer to the actual data of the leaf, but that's different than `it->data`, which is the pointer to the listpack.

there's absolutely no reason for the rax iterator to access the listpack itself.

Reply:
Sorry, i misunderstood, `it->data` is the pointer to the listpack instead of `it->node->data`, please forgive me 😜